File: iot-hub-backend/Domain/RabbitMQ/RPC/RpcRequest.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Domain.RabbitMQ.RPC
{
    public class RpcRequest
    {
        public string ReplyTo { get; set; }
        public string MethodName { get; set; }
        public string ArgumentsJson { get; set; }
        public string CorelationId { get; set; }

        public RpcRequest(string methodName, string argumentsJson, string corelationId, string replyTo)
        {
            ReplyTo = replyTo;
            MethodName = methodName;
            ArgumentsJson = argumentsJson;
            CorelationId = corelationId;
        }

        public byte[] GetJsonBytes()
        {
            var jsonTxt = System.Text.Json.JsonSerializer.Serialize(this);
            return Encoding.UTF8.GetBytes(jsonTxt);
        }
    }
}

File: iot-device/src/Domain/RabbitMQ/RPC/RpcRequest.cpp
#include <iostream>
#include "nlohmann/json.hpp"
using json = nlohmann::json;
using namespace std;

#define Nameof(x) #x

class RpcRequest
{
public:
    string ReplyTo;
    string MethodName;
    string ArgumentsJson;
    string CorelationId;

    RpcRequest(string replyTo, string methodName, string argumentsJson, string corelationId)
    {
        ReplyTo = replyTo;
        MethodName = methodName;
        ArgumentsJson = argumentsJson;
        CorelationId = corelationId;
    }

    RpcRequest(const json &j)
    {
        j.at("ReplyTo").get_to(ReplyTo);
        j.at("MethodName").get_to(MethodName);
        j.at("ArgumentsJson").get_to(ArgumentsJson);
        j.at("CorelationId").get_to(CorelationId);
    }

    json to_json()
    {
        return json{{Nameof(ReplyTo), ReplyTo}, {Nameof(MethodName), MethodName}, {Nameof(ArgumentsJson), ArgumentsJson}, {Nameof(CorelationId), CorelationId}};
    }
};
